File: DDDCore.NET/State.cs
using System.Collections.Generic;

namespace DDD.Domain.Core;

public abstract class State
{
    public void Mutate(IDomainEvent evt)
    {
        ((dynamic)this).When((dynamic)evt);
    }

    public void Replay(IEnumerable<IDomainEvent> domainEvents)
    {
        foreach (var evt in domainEvents)
        {
            Mutate(evt);
        }
    }
}

File: DDDCore.Java/State.java
package com.infosupport.ddd.domain.core;

import java.util.Iterator;

public abstract class State {

    public abstract void Mutate(IDomainEvent evt);
    
    public void Replay(Iterable<IDomainEvent> domainEvents) {
        for (Iterator<IDomainEvent> iterator = domainEvents.iterator(); iterator.hasNext();) {
            Mutate(iterator.next());
        }
    }
}
